src/neo/Ledger/Blockchain.cs

```diff
@@ -294,78 +295,30 @@ private RelayResultReason OnNewBlock(Block block)
         {
             if (block.Index <= Height)
                 return RelayResultReason.AlreadyExists;
-            if (block_cache.ContainsKey(block.Hash))
-                return RelayResultReason.AlreadyExists;
-            if (block.Index - 1 >= header_index.Count)
+            if (block.Index - 1 > Height)
             {
                 AddUnverifiedBlockToCache(block);
                 return RelayResultReason.UnableToVerify;
             }
-            if (block.Index == header_index.Count)
-            {
-                if (!block.Verify(currentSnapshot))
-                    return RelayResultReason.Invalid;
-            }
-            else
-            {
-                if (!block.Hash.Equals(header_index[(int)block.Index]))
-                    return RelayResultReason.Invalid;
-            }
             if (block.Index == Height + 1)
             {
-                Block block_persist = block;
-                List<Block> blocksToPersistList = new List<Block>();
-                while (true)
-                {
-                    blocksToPersistList.Add(block_persist);
-                    if (block_persist.Index + 1 >= header_index.Count) break;
-                    UInt256 hash = header_index[(int)block_persist.Index + 1];
-                    if (!block_cache.TryGetValue(hash, out block_persist)) break;
-                }
-
-                int blocksPersisted = 0;
-                foreach (Block blockToPersist in blocksToPersistList)
+                if (!block.Verify(currentSnapshot))
                 {
-                    block_cache_unverified.Remove(blockToPersist.Index);
-                    Persist(blockToPersist);
-
-                    // 15000 is the default among of seconds per block, while MilliSecondsPerBlock is the current
-                    uint extraBlocks = (15000 - MillisecondsPerBlock) / 1000;
-
-                    if (blocksPersisted++ < blocksToPersistList.Count - (2 + Math.Max(0, extraBlocks))) continue;
-                    // Empirically calibrated for relaying the most recent 2 blocks persisted with 15s network
-                    // Increase in the rate of 1 block per second in configurations with faster blocks
-
-                    if (blockToPersist.Index + 100 >= header_index.Count)
-                        system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = blockToPersist });
+                    system.SyncManager.Tell(new SyncManager.InvalidBlockIndex { InvalidIndex = block.Index });
```

Comment:
Why tell `SyncManager` this message?

Reply:
When Blockchain receive a invalid block, we need to tell SyncManager to retrieve the block of that height and increment the error counter of the node sending the invalid block by 1, which helps us later assess the health of the node.

Reply:
Wait for #1495

Reply:
Please listen to the event stream.

Reply:
https://github.com/neo-project/neo/blob/4755824fece8c577874910fd69702e38ed777193/src/neo/Ledger/Blockchain.cs#L29

Reply:
OK. I will modify it.

Reply:
@erikzhang Done.

Reply:
Then `SyncManager.InvalidBlockIndex` is no more needed.